### PR TITLE
[language] Add move-lang support for hex strings

### PR DESCRIPTION
**Pull request by bob-wilson**
This adds Move support for hexadecimal strings with a syntax of x"...". This is not necessarily the final syntax. The rest of the compiler seems to handle these vector literals already, so only the parser needed to be changed.

## Motivation

Provide the ability to specify literal byte arrays as hexadecimal strings.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I added tests for various corner cases in the lexer and parser, along with a basic functional test.

### CODE REVIEW

File: language/move-lang/src/parser/lexer.rs [hunk 1]
```diff
@@ -252,8 +251,30 @@ fn find_token(file: &'static str, text: &str, start_offset: usize) -> Result<(To
             }
         }
         'A'..='Z' | 'a'..='z' | '_' => {
-            let len = get_name_len(&text);
-            (get_name_token(&text[..len]), len)
+            if text.starts_with("x\"") && text.len() > 2 {
```
Comment (sblackshear) [inline language/move-lang/src/parser/lexer.rs]: Nit: do we want to allow both `xABD` and `XABC` for consistency with the address syntax (which allows both upper and lowercase x)?

I'm happy with only lowercase too (more readable IM), but it seems like whatever we do should be consistent with the address syntax.

File: language/move-lang/tests/move_check/parser/hexstring_bad_value2.exp [hunk 2]
```diff
@@ -0,0 +1,8 @@
+error: 
+
+   ┌── tests/move_check/parser/hexstring_bad_value2.move:3:9 ───
+   │
+ 3 │         x"abcdefg"
+   │         ^^^^^^^^ Missing closing quote (") after hex string
```
Comment (sblackshear) [inline language/move-lang/tests/move_check/parser/hexstring_bad_value2.exp]: This error message looks wrong. Shouldn't it be complaining about the `g`?

File: language/move-lang/src/parser/lexer.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/parser/lexer.rs]: We allow `0XABC`?

File: language/move-lang/tests/move_check/parser/hexstring_bad_value2.exp (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/tests/move_check/parser/hexstring_bad_value2.exp]: Ditto, I feel like we should take the string, and then try to parse it as a hex string. Then give the error.

**Comment by tnowacki** [comment]
>This error message looks wrong. Shouldn't it be complaining about the g?

Due to @sblackshear's error message complaint. I think its actually going to be easier to also add `b""` at the same time. We parse both as strings. Then restrict the encoding as necessary. Then convert to hex. 

The current approach makes that difficult by trying to force into the lexer

**Comment by tnowacki** [comment]
But I'm not going to block the land for that. I think we should do these things, but I don't care too much about the timetable

File: language/move-lang/src/parser/lexer.rs (hunk 1 above)
Comment (bob-wilson) [reply language/move-lang/src/parser/lexer.rs]: I'm neutral on that one. I'll plan to add the uppercase "X" option in a follow-up in no one objects.

File: language/move-lang/tests/move_check/parser/hexstring_bad_value2.exp (hunk 2 above)
Comment (bob-wilson) [reply language/move-lang/tests/move_check/parser/hexstring_bad_value2.exp]: Yeah, I know -- I wasn't very happy with this message, either. I hadn't yet come up with a better solution, but I like Todd's idea.

**Comment by bob-wilson** [comment]
> I think its actually going to be easier to also add `b""` at the same time. We parse both as strings. Then restrict the encoding as necessary. Then convert to hex.

My hesitation on `b""` was twofold: (1) if we decided to go with a completely different syntax for the hex strings, then `b"""` might not make sense, and (2) it seems like a good implementation would need to allow escapes for things like double quotes, newlines, etc. and I didn't want to deal with that.

Another option would be to implement only `x""` (and now `X""`) but move the logic out of the lexer. How sure are we that `b""` will be a useful addition to the language?

I'll go ahead and land this for now but I'll plan to follow up soon to improve this.

**Comment by bob-wilson** [comment]
Hmm. I'm getting lint errors about the missing newlines (intentional) in my EOF tests:
language/move-lang/tests/move_check/parser/hexstring_token_eof1.move: missing a newline at EOF
language/move-lang/tests/move_check/parser/hexstring_token_eof2.move: missing a newline at EOF

Maybe there's a way to override those?

File: language/move-lang/src/parser/lexer.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/parser/lexer.rs]: I kind of don't like the X option. Should we really have it for the 0x numbers? Is it that useful/common/expected?

Comment (bob-wilson) [reply language/move-lang/src/parser/lexer.rs]: C and C++ both allow the uppercase "0X" prefix. I don't think that Rust allows it, though. I don't like uppercase for the actual hex values, either, but I think it's good to allow flexibility on that one. I'd be fine to allow only lowercase "0x" if there is a consensus around that.

Comment (tnowacki) [reply language/move-lang/src/parser/lexer.rs]: I say less is better here. Just 0x seems fine by me

Comment (sblackshear) [reply language/move-lang/src/parser/lexer.rs]: Agreed; I actually prefer lowercase only. My only argument for the uppercase `X` was consistency.

**Comment by bob-wilson** [comment]
I reimplemented this using Todd's suggestion. I didn't actually implement support for b"..." yet, but it would be easier to do that now. This is overall a better way to do it, and it also avoids the need for those "noeol" test files.

**Comment by bob-wilson** [comment]
I'm going to go ahead and land this to unblock Sam's testing, but I welcome any more comments on the revised PR.
